File: kotlin/Lib/787.CheapestFlightsWithinKStops.kt
// 787. Cheapest Flights Within K Stops

data class Edge(val src: Int, val dst: Int, val weight: Int)

class LeetCode787CheapestFlightsWithinKStops {
    fun findCheapestPrice(n: Int, flights: Array<IntArray>, src: Int, dst: Int, K: Int): Int {
        return _FindCheapestPrice(n, flights.map { Edge(it[0], it[1], it[2]) }.toTypedArray(), src, dst, K)
    }

    private fun _FindCheapestPrice(n: Int, edges: Array<Edge>, src: Int, dst: Int, K:Int): Int {
        if (n < 2 || edges.isEmpty()) return 0;

        var dp = IntArray(n) { Int.MAX_VALUE }
        dp[src] = 0

        for (i in 0..K) {
            dp = `_UpdateTable`(edges, dp)
        }

        if (dp[dst] == Int.MAX_VALUE) return -1
        else return dp[dst]
    }

    private fun _UpdateTable(edges: Array<Edge>, dp: IntArray): IntArray {
        val temp = dp.clone()

        for (edge in edges) {
            if (dp[edge.src] == Int.MAX_VALUE) continue

            temp[edge.dst] = minOf(temp[edge.dst], dp[edge.src] + edge.weight)
        }
        return temp
    }
}

File: csharp/src/787_CheapestFlightsWithinKStops.cs
/*
LeetCode: 787. Cheapest Flights Within K Stops
*/

using System;

namespace LeetCode.Problem_787
{
	public class CheapestFlightsWithinKStops {
		private const int FROM = 0;
		private const int TO = 1;
		private const int PRICE = 2;
		public int FindCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
			var table = GenTable_(n);
			table[src] = 0;

			for (int count = 0; count <= K; ++count)
			{
				var tempTable = (int[])table.Clone();
				foreach (var flight in flights)
				{
					var from = flight[FROM];
					if (table[from] == int.MaxValue)
						continue;
					
					var to = flight[TO];
					var price = flight[PRICE];

					tempTable[to] = Math.Min(tempTable[to], table[from] + price);
				}
				table = tempTable;
			}
			return table[dst] == int.MaxValue? -1: table[dst];
		}

		private int[] GenTable_(int numPlaces)
		{
			var table = new int[numPlaces];
			for (int i = 0; i < table.Length; ++i)
				table[i] = int.MaxValue;
			return table;
		}
	}
}